File: 465Assignment1/src/Zpm.java
/**
 * Savannah Lescas
 * Dr. Amjad
 * Comparative Programming Languages
 * 8 February 2024
 * A program that takes in a .zpm file from the command line. In the .zpm file
 * there are Z+- commands. The program takes the commands from the file 
 * and processes/interprets them. 
 */

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Scanner;
import java.io.File;
import java.io.FileNotFoundException;

public class Zpm {
	
	private static HashMap<String, Object> variables = new HashMap<>();
	private static int lineNumber = 0;
	
	public static void main(String[] args){
		// check if a filename is one of the arguments
		if (args.length == 0)  {
			System.out.println("Usage: java Zpm filename.zpm");
			System.exit(1);
		}
		String filename = args[0];
		// make sure that it is a .zpm extension
		if (!filename.endsWith(".zpm")) {
			System.out.println("File needs to have .zpm extension");
			System.exit(1);
		}
		File file = new File(filename);
		try (Scanner scan = new Scanner(file)) {
			// traverse the file
			while (scan.hasNextLine()) {
				String line = scan.nextLine();
				// split the parts of the line into strings
				String[] input = line.split(" ");
				lineNumber++;
				// run it through the doAssignments method
				doAssignments(input);
			}
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		}
	}
	
	/*
	 * A method that gets the first part of the line to determine what it
	 * is trying to do.
	 */
	public static void doAssignments(String[] input) {
		String firstInstruction = input[0];
		// switch case with the first instruction
		switch (firstInstruction) {
			// if the instruction is FOR do the for loop method
			case "FOR":
				loop(input);
				break;
			// if the instruction is PRINT run the print method
			case "PRINT":
				// use the print function to print
				print(input[1]);
				break;
			// if it is not FOR or PRINT it is either a variable assignment
			// or a compound assignment
			default:
			try {
				instruction(input);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}
	
	/*
	 * A method that checks if the line is a variable assignment or a 
	 * compound assignment and does the instruction respectively.
	 */
	public static void instruction(String[] input) {
		// if the second part of the input line is just "=" that means 
		// it is a variable assignment
		if (input[1].equals("=")) {
			// makes sure the line is correctly syntaxed
			if (input.length < 4 || input[2].equals(";")) {
				System.out.println("RUNTIME ERROR: line " + lineNumber);
				System.exit(1);
			} else {
				// check that the variable name has no numbers
				if (checkVarName(input[0])) {
					// if it is all good initialize it and put it into the hash map
					initialize(input[0], input[1], input[2], input[3]);
				} else {
					System.out.println("RUNTIME ERROR: line " + lineNumber);
					System.exit(1);
				}
			}
		// if the second part of the input is not "=", it is a compound assignment
		} else {
			// if the right side of the compound assignment is a variable name,
			// do the assignment with the value from the variable key
			if (variables.containsKey(input[2])) {
				compoundAssignments(input[0], input[1], variables.get(input[2]));
			// if it is not a variable then do the compound assignment regularly
			} else {
				compoundAssignments(input[0], input[1], input[2]);
			}
		}
	}
	
	/*
	 * A method that does the variable initialization. If the variable does not
	 * already exist in the hash map, add it to the map. If the variable is 
	 * getting assigned to another variable's value change the variable's value.
	 */
	public static void initialize(String first, String second,
			String third, String fourth) {
		// if the assignment is to a variable that exists, assign the
		// variable to the value of the other variable
		if (variables.containsKey(third)) {
			Object value = variables.get(third);
			variables.put(first, value);
		} else {
			// create a new key value pair and put it into the hash map
			try {
				Integer intObj = Integer.parseInt(third);
				variables.put(first, intObj);
			} catch (NumberFormatException e) {
				// checking that the assignment is to a string not a non-initialized
				// variable
				if (third.length() >= 2 && third.charAt(0) == '"') {
					String noQuotes = third.substring(1, (third.length() - 1));
					variables.put(first, noQuotes);
				} else {
					// since we know that the length is too short and it isn't
					// any empty string we know that the file is trying to assign
					// a variable to a variable that does not have values so we
					// give a runtime error
					System.out.println("RUNTIME ERROR: line " + lineNumber);
					System.exit(1);
					
				}
			}
		}
	}
	
	/*
	 * A method that does the compound assignments. 
	 */
	public static void compoundAssignments(String variableName,
			String operator, Object right) {
		// checks if the variable has been initialized yet
		if (variables.get(variableName) != null) {
			String left = variables.get(variableName).toString();
			// if both sides are strings concatenate them
			if (!isInt(left) && !isInt(right.toString())) {
				if (operator.equals("+=")) {
					variables.put(variableName, String.valueOf(left) + String.valueOf(right));
				}
			// if the command is trying to add a string with an int
			// it will be a runtime error
			} else if (!isInt(left) && isInt(right.toString())
					|| isInt(left) && !isInt(right.toString())) {
				System.out.println("RUNTIME ERROR: line " + lineNumber);
				System.exit(1);
			// if the compound assignment is with two integers, do the assignment
			// and update the value in the map
			} else {
				int leftInt = Integer.parseInt(left);
				int rightInt = Integer.parseInt(right.toString());
				// figure out which compound assignment it is doing
				if (operator.equals("+=")) {
					variables.put(variableName, leftInt + rightInt);
				} else if (operator.equals("*=")) {
					variables.put(variableName, leftInt * rightInt);
				} else if (operator.equals("-=")) {
					variables.put(variableName, leftInt - rightInt);
				} else {
					System.out.println("RUNTIME ERROR: line " + lineNumber);
					System.exit(1);
				}
			}
		// if the variable is not initialized already, give a runtime error
		} else {
			System.out.println("RUNTIME ERROR: line " + lineNumber);
			System.exit(1);
		}
	}
	
	/*
	 * A method that prints the variable if it exists.
	 */
	public static void print(String variableName) {
		if (variables.get(variableName) != null) {
			System.out.println(variableName + "=" + variables.get(variableName));
		} else {
			System.out.println("RUNTIME ERROR: line " + lineNumber);
			System.exit(1);
		}
		
	}
	
	/*
	 * A method that performs the for loop. 
	 */
	public static void loop(String[] input) {
		// change the iterations string into an int
		int iterations = Integer.parseInt(input[1]);
		// an array list to hold the statements as if they were on their
		// own line
		ArrayList<String> statements = new ArrayList<String>();
		String statement = "";
		for (int i = 2; i < input.length; i++) {
			if (input[i].trim().equals("ENDFOR")){
				// use the iterations from the line to determine how many times
				// to do the loop
				for (int j = 0; j < iterations; j++) {
					// another loop to go through the array list that has each line
					// to repeat
					for (int k = 0; k < statements.size(); k++) {
						// take each task and separate it so the doAssignments method
						// can do the line
						String[] instruction = statements.get(k).split(" ");
						doAssignments(instruction);
					}
				}
			} else if (!input[i].equals(";")) {
				statement += input[i] + " ";
			} else {
				statements.add(statement);
				statement = "";
			}
		}
	}
	
	/*
	 * A method that checks to make sure the variable name the file is
	 * trying to use only letters.
	 * Returns true if the name is only letters and false otherwise.
	 */
	private static boolean checkVarName(String varName) {
		for (int i = 0; i < varName.length(); i++) {
			if (!(varName.charAt(i) >= 'A' && varName.charAt(i) <= 'Z') 
					|| varName.length() > 1) {
				return false;
			}
		}
		return true;
	}
	
	/*
	 * A method that checks if the input is an int.
	 */
	private static boolean isInt(String input) {
		//returns true if everything in string is a number
		return input.trim().matches("-?(0|[1-9]\\d*)");
	}
 }

File: Homework6/Zpm.cpp
// Copyright 2024
// Savannah Lescas
#include <iostream>
#include <unordered_map>
#include <vector>
#include <fstream>
#include <sstream>
#include <regex>

using namespace std;
void doAssignments(vector<string> input);

unordered_map<string, string> variables;
int lineNumber = 0;

bool isInt(string input) {
    regex pattern("-?(0|[1-9]\\d*)");

    return regex_match(input, pattern);
}

bool checkVarName(string varName) {
    regex pattern("^[a-zA-Z][\\w$]*$");

    return regex_match(varName, pattern);

}


string getRidOfQuotes(string quoted) {
    string noQuotes = "";
    for (size_t i = 0; i < quoted.length(); i++) {
        if (quoted[i] != '"') {
            noQuotes += quoted[i];
        }
    }
    return noQuotes;
}

void print(string variableName) {
    auto search = variables.find(variableName);
    if (search != variables.end()) {
        cout << variableName << "=" << search->second << endl;
    } else {
        cout << "RUNTIME ERROR: line " << lineNumber << endl;
        exit(1);
    }
}

void initialize(string first, string second, string third, string fourth) {
    auto search = variables.find(third);
    if (search != variables.end()) {
        string value = search->second;
        variables[first] = value;
    } else {
        variables[first] = third;
    }
}

void compoundAssignments(string variableName, string op, string right) {
    auto search = variables.find(variableName);
    if (search != variables.end()) {
        string left = getRidOfQuotes(search->second);
        if (!isInt(left) && !isInt(right)) {
            if (op == "+=") {
                variables[variableName] = '"' + left + right + '"';
            }
        } else if (!isInt(left) && isInt(right)
            || isInt(left) && !isInt(right)) {
                cout << "RUNTIME ERROR: line " << lineNumber << endl;
                exit(1);
        } else {
            int leftInt = stoi(left);
            int rightInt = stoi(right);
            if (op == "+=") {
                string total = to_string(leftInt + rightInt);
                variables[variableName] = total;
            } else if (op == "-=") {
                string total = to_string(leftInt - rightInt);
                variables[variableName] = total;
            } else if (op == "*=") {
                string total = to_string(leftInt * rightInt);
                variables[variableName] = total;
            } else {
                cout << "RUNTIME ERROR: line " << lineNumber << endl;
                exit(1);
            }

        }
    } else {
        cout << "RUNTIME ERROR: line " << lineNumber << endl;
        exit(1);
    }
}

void stringCheck(vector<string>& input) {
    string newString = "";
    int index;
    auto search = variables.find(input[2]);
    if (!isInt(input[2]) && search == variables.end()) {
        // a condition that checks if two elements from the input vector
        // next to each other are both quotes, if they are then that
        // means the variable was meant to be assigned as a space
        if (input.size() > 3) {
            if (input[2][0] == '"' && input[3][0] == '"') {
                newString += " ";
            }
        }
        for (size_t i = 2; i < input.size(); i++) {
            // if it starts with a quote and does not end with 
            // a quote then it is part of a longer string
            if (input[i][0] == '"' && input[i].back() != '"') {
                index = i;
                // so add it to the new string and add a space
                newString += input[i] + " ";
                break;
            } else {
                // otherwise, the string is just one word and the 
                // space isn't necessary
                index = i;
                //cout << "input[i]: " << input[i] << endl;
                newString += input[i] + "";
                break;
            }
        }
        for (size_t i = index + 1; i < input.size(); i++) {
            if (input[i] == ";") {
                break;
            } else if (!(input[i].back() == '"')) {
                newString += input[i] + " ";
                continue;
            } else {
                newString += input[i];
                break;
            }
        }
    } else {
        newString += input[2];
    }
    input[2] = newString;
    if (input.size() >= 4) {
        input[3] = ";";
        while (input.size() > 4) {
            input.pop_back();
        }   
    } else {
        input.push_back(";");
    }
}


void instruction(vector<string>& input) {
    stringCheck(input);
    if (input[1] == "=") {
        // this checks that the line has at least 4 parts and that the
        // last part is a semicolon
        if (input.size() < 4 || !(input[input.size() - 1] == ";")) {
            cout << "first check RUNTIME ERROR: line " << lineNumber << endl;
            exit(1);
        } else {
            if (checkVarName(input[0])) {
                initialize(input[0], input[1], input[2], input[3]);
            } else {
                cout << "RUNTIME ERROR: line " << lineNumber << endl;
                exit(1);
            }
        }
    } else {
        // if compounding a variable with an existing variable
        auto search = variables.find(input[2]);
        if (search != variables.end()) {
            // get the value of the variable
            string value = search->second;
            compoundAssignments(input[0], input[1], getRidOfQuotes(value));
        } else {
            compoundAssignments(input[0], input[1], getRidOfQuotes(input[2]));
        }
    }
}

void loop(vector<string> input) {
    int iterations;
    auto search = variables.find(input[1]);
    if (search != variables.end()){
        iterations = stoi(search->second);
    } else {
        iterations = stoi(input[1]);
    }


    vector<string> statements = {};
    string statement = "";
    for (int i = 2; i < input.size(); i++) {
        if (input[i] == "ENDFOR") {
            for (int j = 0; j < iterations; j++) {
                for (int k = 0; k < statements.size(); k++) {
                    vector<string> instruction;
                    string s;
                    stringstream ss(statements[k]);
                    while (getline(ss, s, ' ')) {
                        instruction.push_back(s);
                    }
                    doAssignments(instruction);
                }
            }
        } else if (input[i] != ";") {
            statement += input[i] + " ";
        } else {
            statements.push_back(statement);
            statement = "";
        }
    }
}

void doAssignments(vector<string> input) {
    string firstInstruction = input[0];

    if (firstInstruction == "FOR") {
        loop(input);
    } else if (firstInstruction == "PRINT") {
        print(input[1]);
    } else {
        try {
            instruction(input);
        } catch (...) {
            cout << "Error" << endl;
        }
    }
}



int main (int argc, char *argv[]) {
    string filename = "";
    if(argc == 2) {
        filename = argv[1];
    } else {
        cout << "Usage: ./Zpm filename.zpm" << endl;
        exit(1);
    }

    regex regex(".*\\.zpm");
    // making sure file ends with ".zpm"
    if (!regex_match(filename, regex)) {
        cout << "File needs to have .zpm extension" << endl;
        exit(1);
    }

    ifstream inputFile(filename);
    if (!inputFile) {
        cout << "File could not be opened" << endl;
        exit(1);
    }

    string line;
    vector<string> input;
    while (!inputFile.eof()) {
        getline(inputFile, line);
        lineNumber++;
        if (line.empty()) {
            continue;
        }
        input.clear();
        string s;
        stringstream ss(line);
        while (getline(ss, s, ' ')) {
            input.push_back(s);
        }

        doAssignments(input);
    }

    return (0);
}